### ISSUE TITLE
PaxosTimestampBoundStore - Infinite Loop of Network Requests

### ISSUE THREAD

**Issue by jeremyk-91** [palantir/atlasdb#1941]
Internal reference is PDS-52730.

We established that under network strain, it is possible for a node to lose leadership and then regain it quickly, so we can have multiple running timestamp services on the same box. This happened at internal deployment.

Investigating the request logs, it was apparent that there was a successful call at sequence 98393 at some point. However, this raced with another ongoing update. Consider the code of storeUpperLimit():

```java
    @Override
    public synchronized void storeUpperLimit(long limit) throws MultipleRunningTimestampServiceError {
        long newSeq = PaxosAcceptor.NO_LOG_ENTRY + 1;
        if (agreedState != null) {
            Preconditions.checkArgument(limit >= agreedState.getBound(),
                    "Tried to store an upper limit %s less than the current limit %s", limit, agreedState.getBound());
            newSeq = agreedState.getSeqId() + 1;
        }
        while (true) {
            try {
                proposer.propose(newSeq, PtBytes.toBytes(limit));
                PaxosValue value = knowledge.getLearnedValue(newSeq);
                checkAgreedBoundIsOurs(limit, newSeq, value);
                long newLimit = PtBytes.toLong(value.getData());
                agreedState = ImmutableSequenceAndBound.of(newSeq, newLimit);
                if (newLimit >= limit) {
                    return;
                } // but if not this will NEVER succeed.
            } catch (PaxosRoundFailureException e) {
                waitForRandomBackoff(e, this::wait);
            }
        }
    }
```

Because we reuse Paxos proposers across multiple instantiations of the timestamp service on the same node, checkAgreedBoundIsOurs will pass if the bound was set by another timestamp service on the same node. So this trace is possible (note: I don't have enough evidence to confirm this is what happened, though the symptoms very closely match):

1. A is the only timestamp service, and writes a value (X, 98392).
2. A wants to store (X + 500K, 98393).
3. Leader elections occur, and B on the same node wants to store (X + 1M, 98393).
4. A's proposal wins, so we have (X + 500K, 98393)
5. B's proposal fails since we already have a 98393.
6. B tries to propose (X + 1M, 98393). Of course, this is going to return (X + 500K, 98393).
7. Repeat from step 5, ad infinitum.

Since these steps can happen very quickly we end up spawning many network requests quickly, and re-proposing at the same sequence is never going to succeed - we get an infinite loop. In this case we need to be very careful. Re-proposing from an increased sequence number will mean that whoever is behind us might issue duplicate timestamps.

_I think as a first fix, it is acceptable for us to throw NCLE in this case_, forcing a leadership election and invalidating the timestamp store. This is a P1 pain for our customer and that will (I think) safely resolve it; we can then later consider ways of fixing this that preserve greater availability.

@jamding @gsheasby for SA

**Comment by nziebart** [comment]
Does throwing a NCLE actually trigger a re-election / invalidation of the timestamp service? I was under the impression that it just returns a 503 to the caller. 

I'm not sure if this is what's causing this particular issue, but re-using the same proposer UUID across elections definitely feels like a bug. In theory, you could have two threads on the same node try to perform the same CAS (e.g., from 1,000 to 2,000), and both succeed - in which case they would give out duplicate timestamps.

I think we should fix the `PaxosTimestampBoundStore` so that the CAS guarantees hold even if the two callers are on the same node. In practice I think this could be done by (a) creating a new proposer UUID for each election or (b) generate a UUID for each CAS request, append it to the paxos value, and check that rather than the proposer UUID (i.e., ignore the proposer UUID).

**Comment by jeremyk-91** [comment]
Yeah, `AwaitingLeadershipProxy` clears out the delegate if a NCLE or ServiceNotAvailableException occurs (though as we've discussed offline there are issues with the way we threw NCLE in 1942, which are fixed in 0.42.2 and should be making their way to develop tomorrow).

The reusing UUID two-successful-CASes is filed at #1937 if I understand correctly. WIth the async mechanism we had this is unlikely, but I think after we switch to sync update in #1910 most updates _are_ going to be 1M, which could increase the risk this might occur.

I'd be fine with either of the solutions; I prefer (a) assuming that we don't make assumptions anywhere that that is constant across election terms. In terms of debugging information which is a concern, the timestamp bound store starts up by logging its proposer ID, so we don't lose anything (well, apart from cases where logs are partially rolled).

**Comment by nziebart** [comment]
Agree that (a) feels like the best solution - it avoids changing the paxos value format.

**Comment by hsaraogi** [comment]
fixed by #1942.
